pkg/cluster/check/discovery.go

```diff
@@ -0,0 +1,102 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package check
+
+import (
+	"net/netip"
+
+	"inet.af/netaddr"
+
+	"github.com/talos-systems/talos/pkg/cluster"
+	"github.com/talos-systems/talos/pkg/machinery/config/types/v1alpha1/machine"
+	"github.com/talos-systems/talos/pkg/machinery/generic/maps"
+	"github.com/talos-systems/talos/pkg/machinery/generic/slices"
+	clussterres "github.com/talos-systems/talos/pkg/machinery/resources/cluster"
+)
+
+// DiscoveredClusterInfo represents a cluster.Info populated using the discovery service.
+type DiscoveredClusterInfo struct {
+	nodes       []cluster.NodeInfo
+	nodesByType map[machine.Type][]cluster.NodeInfo
+}
+
+// Nodes returns list of all node infos.
+func (d *DiscoveredClusterInfo) Nodes() []cluster.NodeInfo {
+	return d.nodes
+}
+
+// NodesByType return list of node endpoints by type.
+func (d *DiscoveredClusterInfo) NodesByType(m machine.Type) []cluster.NodeInfo {
+	return d.nodesByType[m]
+}
+
+// NewDiscoveredClusterInfo returns a new cluster.Info populated from the discovery service.
+func NewDiscoveredClusterInfo(members []clussterres.Member) (cluster.Info, error) {
+	m, err := membersToNodeInfoMap(members)
+	if err != nil {
+		return nil, err
+	}
+
+	nodes := slices.FlatMap(maps.Values(m), identity[[]cluster.NodeInfo])
```

Comment:
I think we can use lambda here?

Reply:
you mean just inlining it as func?

Reply:
Yup

Reply:
Done (wish go had arrow syntax e.g. `t -> t`)